pallets/contracts2/src/lib.rs

```diff
@@ -0,0 +1,438 @@
+    let caller_did = Identity::<T>::key_to_identity_dids(key);
+    Context::set_current_identity::<Identity<T>>(Some(caller_did));
+    old_did
+}
+
+fn decode<V: Decode, T: Config>(input: &mut &[u8]) -> Result<V, DispatchError> {
+    <_>::decode(input).map_err(|_| pallet_contracts::Error::<T>::DecodingFailed.into())
+}
+
+fn construct_call<T>(func_id: FuncId, input: &mut &[u8]) -> Result<CommonCall<T>, DispatchError>
+where
+    T: Config + pallet_asset::Config,
+{
+    /// Decode type from `input`.
+    macro_rules! decode {
+        () => {
+            decode::<_, T>(input)?
+        };
+    }
+
+    /// Pattern match on functions `0x00_pp_ee_00`.
+    macro_rules! on {
+        ($p:pat, $e:pat) => {
+            FuncId {
+                scheme: 0,
+                pallet: $p,
+                extrinsic: $e,
+                version: 0,
+            }
+        };
+    }
+
+    // TODO(Centril): Should we charge weight depending on branch?
+    // That would require us to benchmark the decoding of each and every call.
+
+    Ok(match func_id {
+        on!(0, 0) => CommonCall::Asset(pallet_asset::Call::register_ticker { ticker: decode!() }),
+        on!(0, 1) => {
+            CommonCall::Asset(pallet_asset::Call::accept_ticker_transfer { auth_id: decode!() })
+        }
+        on!(0, 2) => CommonCall::Asset(pallet_asset::Call::accept_asset_ownership_transfer {
+            auth_id: decode!(),
+        }),
+        on!(0, 3) => CommonCall::Asset(pallet_asset::Call::create_asset {
+            name: decode!(),
+            ticker: decode!(),
+            divisible: decode!(),
+            asset_type: decode!(),
+            identifiers: decode!(),
+            funding_round: decode!(),
+            disable_iu: decode!(),
+        }),
+        _ => return Err(Error::<T>::RuntimeCallNotFound.into()),
+    })
+}
+
+/// A chain extension allowing calls to polymesh pallets
+/// and using the contract's DID instead of the caller's DID.
+impl<T> ce::ChainExtension<T> for Module<T>
+where
+    <T as pallet_contracts::Config>::Call: From<CommonCall<T>> + GetDispatchInfo,
+    T: Config + pallet_asset::Config,
+{
+    fn enabled() -> bool {
+        true
+    }
+
+    fn call<E: ce::Ext<T = T>>(
+        func_id: u32,
+        env: ce::Environment<E, ce::InitState>,
+    ) -> ce::Result<ce::RetVal> {
+        let mut env = env.buf_in_buf_out();
+
+        // TODO(Centril): charge weight + benchmark code in here?
+        // It becomes a bit cyclical,
+        // as charging weight itself becomes part of the computation to charge weight for.
+
+        // Swap current DID to caller's DID.
+        let old_did = set_current_did_to_key::<T>(env.ext().caller());
+
+        // Decide what to call in the runtime.
+        let func_id = split_func_id(func_id);
+        // TODO(Centril): charge weight + benchmark depending on `in_len`.
+        // Also, should we impose a max limit on `in_len`?
```

Comment:
Thoughts?

Reply:
I think having a limit on `in_len` is a good idea.

It might be best to use a per-FuncId limit.  Some FuncIds have fixed size parameters and `in_len` could be checked against that before even trying to decode the parameters.